tutor/hooks/consts.py

```diff
@@ -0,0 +1,266 @@
+"""
+List of all the action, filter and context names used across Tutor. This module is used
+to generate part of the reference documentation.
+"""
+# The Tutor plugin system is licensed under the terms of the Apache 2.0 license.
+__license__ = "Apache 2.0"
+
+from tutor import hooks
+
+__all__ = ["Actions", "Filters", "Contexts"]
+
+
+class Actions:
+    """
+    This class is a container for the names of all actions used across Tutor
+    (see :py:mod:`tutor.hooks.actions.do`). For each action, we describe the
+    arguments that are passed to the callback functions.
+
+    To create a new callback for an existing action, write the following::
+
+        @hooks.Actions.YOUR_ACTION.add()
+        def your_action():
+            # Do stuff here
+    """
+
+    #: Called whenever the core project is ready to run. This is the right time to install plugins, for instance.
+    #:
+    #: This action does not have any parameter.
+    CORE_READY = hooks.actions.get("core:ready")
+
+    #: Called as soon as we have access to the Tutor project root.
+    #:
+    #: :parameter str root: absolute path to the project root.
+    CORE_ROOT_READY = hooks.actions.get("core:root:ready")
+
+    #: This action is done to auto-detect plugins. In particular, we auto-discover the following plugins:
+    #:
+    #:   - Python packages that declare a "tutor.plugin.v0" entrypoint.
+    #:   - Python packages that declare a "tutor.plugin.v1" entrypoint.
+    #:   - YAML and Python plugins stored in ~/.local/share/tutor-plugins (as indicated by ``tutor plugins printroot``)
+    #:   - When running the binary version of Tutor, official plugins that ship with the binary are automatically discovered.
+    #:
+    #: Discovering a plugin is typically done by the Tutor plugin mechanism. Thus, plugin
+    #: developers probably don't have to implement this action themselves.
+    #:
+    #: This action does not have any parameter.
+    DISCOVER_PLUGINS = hooks.actions.get("plugins:discover")
+
+    #: Load a specific plugin. Only plugins that have previously been discovered can be loaded.
+    #: (see :py:data:`DISCOVER_PLUGINS`)
+    #:
+    #: Plugins are typically loaded because they are enabled.
+    #:
+    #: Most plugin developers will not have to implement this action themselves, unless
+    #: they want to perform a specific action at the moment the plugin is enabled.
+    #:
+    #: This action does not have any parameter.
+    LOAD_PLUGIN = hooks.actions.get_template("plugins:load:{0}")
+
+    #: Load a list of plugins. This action triggers :py:data:`LOAD_PLUGIN` for every
+    #: loaded plugin. Plugins are loaded in alphabetical order.
+    #:
+    #: :parameter list[str] names: list of plugin names.
+    LOAD_PLUGINS = hooks.actions.get("plugins:load")
+
+    #: Triggered after all plugins have been loaded.
+    #:
+    #: This action does not have any parameter.
+    LOAD_PLUGINS_POST = hooks.actions.get("plugins:load-post")
+
+
+class Filters:
```

Comment:
I like the names of all of these filters except `config:base` (see my comment about config in the plugin tutorial) and the `app:tasks:*...` filters. My litmus test here is: if I can look at the name of the filter and reasonably guess what it means and its type signature *without* looking at the docs, then it's a solid name.

I don't have any better suggestions yet... in fairness, the `app:tasks:*` filters are certainly the most complex ones. I need to submit this review soon so I can join a meeting, but I will stew on this and let you know if I have ideas :)

Reply:
Hmmm I think I understand what's bothering you with tasks. Tasks designate both image building/pushing/etc. jobs and scripts that are run as part of platform initialisation:

    APP_TASKS_IMAGES_PULL = hooks.filters.get("app:tasks:images:pull")
    APP_TASKS_INIT = hooks.filters.get("app:tasks:init")

Also, the "app:" namespace does not really make sense because we never use it without the "tasks:" namespace.

The "task" prefix is also not appropriate, because the filters actually yield paths to templates that includes commands.

I suggest the following renaming:

- app:tasks:images:* -> images:*
- app:tasks:init -> commands:init
- app:tasks:pre-init -> commands:pre-init

Reply:
@regisb yes, I think you captured exactly what was bothering me!

I was in the process of typing a comment to suggest:
* app:tasks:images:* -> app:images:*
* app:tasks:init -> app:lifecycle:init-scripts
* app:tasks:pre-init -> app:lifecycle:pre-init-scripts

I like your suggestions as well.

The other thing I had on my mind was the type signatures of these filters. Some of them, especially the build filter, are pretty unwieldy to use:

```python
import typing as t
from tutor import hooks, types
@hooks.Filters.APP_TASK_IMAGES_BUILD.add()
def _add_myplugin_image_to_build(
    build_images: t.List[t.Tuple[str, t.Tuple[str, ...], str, t.List[str]]],
    config: types.Config,
) -> t.List[t.Tuple[str, t.Tuple[str, ...], str, t.List[str]]]:
    ...
```

In addition to modifying the names, friendlier type signature might make these easier to use:
```python
from tutor import hooks, types
@hooks.Filters.IMAGES_TO_BUILD.add()
def _add_myplugin_image_to_build(
    build_images: t.List[types.ImageToBuild], config: types.Config
) -> t.List[types.ImageToBuild]:
    ...
```

where `ImageToBuild` could just be a simple namedtuple or dataclass.

Reply:
I must admit that I did not think about automated type checking for plugin developers... In an ideal world, I would like to create the `IMAGES_BUILD` filter as follows:

    IMAGES_BUILD: SomeFilterType = filters.get("...")

And then developers would write:

    IMAGES_BUILD.add_item(my_item)

and the type of `my_item` would automatically be checked against `SomeFilterType`. 

Would that resolve your issue? Do you know how to implement `SomeFilterType`? Hmmm I'm guessing that I should be using [generics](https://mypy.readthedocs.io/en/stable/generics.html)... At the moment I'm confused how to implement this as long as [this fix](https://github.com/python/mypy/issues/11833) was not added to a release. May I suggest that we wait before we type-annotate all filters and actions?

Reply:
Ack, that mypy bug is such a bummer! Otherwise, this would be a perfect use case for generics:

```python



# tutor/hooks/filters.py
T = t.TypeVar("T")
P = t.TypeVar("P", bound=tuple)
class Filter(Generic[T, P]):
    ....
    def get(name: str, return_type: type[T], param_types: type[P]): Filter[T, P]:
        ....
    def add(self) -> Callable[[CallableFilter[T, P]], CallableFilter[T, P]]:
        ....

# tutor/hooks/const.py
class ImageBuild(NamedTuple):
     service: str
     path: list[str]
     image: str
     args: list
class Filters:
    ....
    IMAGES_BUILD = filters.get("images:build", [list[ImageBuild], [Config]])

# myplugin.py
@hooks.Filters.IMAGES_BUILD.add()
def _add_myplugin_image_to_build(build_images, config):
    # Since IMAGES_BUILD is fully annotated, mypy should be able
    # to typecheck this function without the user having to annotate anything!
    ....  
```
Alas. I'm okay with holding off on annotations until we're able to absorb that mypy bugfix.